packages/pluggableWidgets/heatmap-chart-web/src/HeatMap.tsx

```diff
@@ -0,0 +1,112 @@
+import classNames from "classnames";
+import { createElement, ReactElement, useMemo } from "react";
+import { ChartWidget, ChartWidgetProps } from "@mendix/shared-charts";
+import { HeatMapContainerProps } from "../typings/HeatMapProps";
+import { useHeatMapDataSeries } from "./hooks/data";
+import { createHeatMapAnnotation } from "./utils/annotation";
+
+type HeatMapAnnotation = ReturnType<typeof createHeatMapAnnotation>;
+
+const heatmapChartLayoutOptions: ChartWidgetProps["layoutOptions"] = {
+    font: {
+        color: "#555",
+        size: 12
+    },
+    legend: {
+        font: {
+            family: "Open Sans",
+            size: 14
+        }
+    },
+    xaxis: {
+        fixedrange: true,
+        ticks: ""
+    },
+    yaxis: {
+        fixedrange: true,
+        ticks: ""
+    }
+};
+const heatmapChartConfigOptions: ChartWidgetProps["configOptions"] = {
+    responsive: true
+};
+const heatmapChartSeriesOptions: ChartWidgetProps["seriesOptions"] = {
+    type: "heatmap",
+    hoverinfo: "none",
+    xgap: 1,
+    ygap: 1,
+    colorbar: {
+        y: 1,
+        yanchor: "top",
+        ypad: 0,
+        xpad: 5,
+        outlinecolor: "#9ba492"
+    },
+    connectgaps: false
+};
+
+export function HeatMap(props: HeatMapContainerProps): ReactElement | null {
+    const heatmapChartData = useHeatMapDataSeries({
+        customSeriesOptions: props.customSeriesOptions,
+        horizontalAxisAttribute: props.horizontalAxisAttribute,
+        horizontalSortAttribute: props.horizontalSortAttribute,
+        horizontalSortOrder: props.horizontalSortOrder,
+        onClickAction: props.onClickAction,
+        scaleColors: props.scaleColors,
+        seriesDataSource: props.seriesDataSource,
+        seriesName: props.seriesName,
+        seriesValueAttribute: props.seriesValueAttribute,
+        showScale: props.showScale,
+        smoothColor: props.smoothColor,
+        tooltipHoverText: props.tooltipHoverText,
+        verticalAxisAttribute: props.verticalAxisAttribute,
+        verticalSortAttribute: props.verticalSortAttribute,
+        verticalSortOrder: props.verticalSortOrder
+    });
+
+    const heatmapChartLayout = useMemo<ChartWidgetProps["layoutOptions"]>(
+        () => ({
+            ...heatmapChartLayoutOptions,
+            annotations: props.showValues
+                ? heatmapChartData[0].z.reduce<HeatMapAnnotation[]>(
```

Comment:
What is data is empty?

Reply:
Based on [the return value](https://github.com/mendix/widgets-resources/pull/1234/files#diff-8f234248ed7a972a957d8549363c5a66a08fe89618e0c05de7617c5b413569a7R138-R154) of the `useHeatMapDataSeries` hook, that's not possible. There will always be an array with 1 item